golemfactory/yagna
# Payment platform negotiation

Payment service will now use negotiated payment platform instead of hard-coded default. It is encoded in `golem.com.payment.chosen-platform` property of the final negotiated demand.

Provider and requestor (mock) agents have been modified to specify supported platforms by `golem.com.payment.platform.<name>.address` properties. Mock requestor additionally puts constraints on these properties to filter out offers without appropriate platforms.

The change has been implemented in backwards-compatible manner so that requestors which don't specify payment platform would be treated as using the default platform (i.e. NGNT).

----
Testing instructions:

1. Start the yagna daemon with both NGNT (default) and dummy driver.
```
$ cargo run --features dummy-driver service run
```

2. Create 3 identities: one for requestor and two for providers.
```
$ cargo run id create provider1
[...]
$ cargo run id create provider2
[...]
$ cargo run id create requestor
[...]
$ cargo run id list
┌───────────┬──────────┬─────────────┬──────────────────────────────────────────────┐
│  default  │  locked  │  alias      │  address                                     │
├───────────┼──────────┼─────────────┼──────────────────────────────────────────────┤
│  X        │          │             │  0x5868e603af883e4d2c86940c571030723f8ebca4  │
│           │          │  requestor  │  0x2f9684374f49d3b55ea5907024d34ea7ae0c3055  │
│           │          │  provider2  │  0x5e9f8ab7b4816f527d988dcc0887b30499749ca2  │
│           │          │  provider1  │  0x2aef55f45ef7e911c2d63a1b39d2be572c8d3830  │
└───────────┴──────────┴─────────────┴──────────────────────────────────────────────┘
```

3. Prepare `accounts.json` so that provider1 supports NGNT driver, provider2 supports dummy, and requestor supports both.
```
$ cat accounts.json 
[
  {"driver":"ngnt","address":"0x2aef55f45ef7e911c2d63a1b39d2be572c8d3830","send":false,"receive":true},
  {"driver":"dummy","address":"0x5e9f8ab7b4816f527d988dcc0887b30499749ca2","send":false,"receive":true},
  {"driver":"ngnt","address":"0x2f9684374f49d3b55ea5907024d34ea7ae0c3055","send":true,"receive":false},
  {"driver":"dummy","address":"0x2f9684374f49d3b55ea5907024d34ea7ae0c3055","send":true,"receive":false}
]
```

4. Restart daemon.

5. Verify the accounts are properly initialized.
```
$ cargo run payment accounts

┌────────────┬──────────────────────────────────────────────┬──────────┬────────┬────────┐
│  platform  │  address                                     │  driver  │  send  │  recv  │
├────────────┼──────────────────────────────────────────────┼──────────┼────────┼────────┤
│  DUMMY     │  0x2f9684374f49d3b55ea5907024d34ea7ae0c3055  │  dummy   │  X     │        │
│  DUMMY     │  0x5e9f8ab7b4816f527d988dcc0887b30499749ca2  │  dummy   │        │  X     │
│  NGNT      │  0x2aef55f45ef7e911c2d63a1b39d2be572c8d3830  │  ngnt    │        │  X     │
│  NGNT      │  0x2f9684374f49d3b55ea5907024d34ea7ae0c3055  │  ngnt    │  X     │        │
└────────────┴──────────────────────────────────────────────┴──────────┴────────┴────────┘
```

6. Create app keys for all identities.
```
$ cargo run app-key create provider1 --id provider1
[...]
$ cargo run app-key create provider2 --id provider2
[...]
$ cargo run app-key create requestor --id requestor
[...]
$ cargo run app-key list
┌─────────────┬────────────────────────────────────┬──────────────────────────────────────────────┬───────────┬─────────────────────────────────┐
│  name       │  key                               │  id                                          │  role     │  created                        │
├─────────────┼────────────────────────────────────┼──────────────────────────────────────────────┼───────────┼─────────────────────────────────┤
│  provider1  │  3808b6921666456389139457a0f4cfad  │  0x2aef55f45ef7e911c2d63a1b39d2be572c8d3830  │  manager  │  2020-09-28T15:03:54.821193554  │
│  provider2  │  c62d6a68d50849d29a6ab605d8675a90  │  0x5e9f8ab7b4816f527d988dcc0887b30499749ca2  │  manager  │  2020-09-28T15:06:14.173808111  │
│  requestor  │  c4ac4e04b2c643c4bfec1ab4e534374e  │  0x2f9684374f49d3b55ea5907024d34ea7ae0c3055  │  manager  │  2020-09-28T15:08:17.087101304  │
└─────────────┴────────────────────────────────────┴──────────────────────────────────────────────┴───────────┴─────────────────────────────────┘
```

7. Start two providers (setting the appropriate app keys).
```
$ cargo run -p ya-provider run
```

8. Start the requestor.
```
$ cargo run -p ya-requestor -- --exe-script ../exe-unit/examples/commands.json
```

9. Verify that provider1 is getting paid in NGNT and provider2 is getting paid in dummies.
```
$ cargo run payment status -p dummy 0x5e9f8ab7b4816f527d988dcc0887b30499749ca2
---
amount: "1000000000000000000000000"
incoming:
  accepted: "100.11060284660000"
  confirmed: "100.11060284660000"
  requested: "140.97986801310000"
[...]
$ cargo run payment status -p ngnt 0x2aef55f45ef7e911c2d63a1b39d2be572c8d3830
---
amount: "282.3430044842"
incoming:
  accepted: "90.92201856350000"
  confirmed: "90.92201856350000"
  requested: "105.05616664480000"
[...]
$ cargo run payment status -p ngnt 0x2f9684374f49d3b55ea5907024d34ea7ae0c3055
---
amount: "1403.172907505"
[...]
outgoing:
  accepted: "90.92201856350000"
  confirmed: "90.92201856350000"
  requested: "105.05616664480000"
reserved: "509.07798143650000"
$ cargo run payment status -p dummy 0x2f9684374f49d3b55ea5907024d34ea7ae0c3055
---
amount: "1000000000000000000000000"
[...]
outgoing:
  accepted: "100.11060284660000"
  confirmed: "100.11060284660000"
  requested: "140.97986801310000"
reserved: "499.88939715340000"
```
